File: iOS Car Dealer Tracker/Ezcar24Business/Services/Calculators/HoldingCostCalculator.swift
import Foundation

class HoldingCostCalculator {
    
    private let settings: HoldingCostSettings?
    private static let eligibleStatuses: Set<String> = [
        "in_transit",
        "on_sale",
        "available",
        "reserved",
        "under_service",
        "owned",
        "sold"
    ]
    
    init(settings: HoldingCostSettings? = nil) {
        self.settings = settings
    }
    
    func calculateHoldingCost(for vehicle: Vehicle) -> Decimal {
        guard let settings = settings else { return 0 }
        guard Self.isHoldingCostEligible(vehicle: vehicle) else { return 0 }
        
        let expenses = vehicle.expenses?.allObjects as? [Expense] ?? []
        return Self.calculateAccumulatedHoldingCost(
            vehicle: vehicle,
            settings: settings,
            allExpenses: expenses
        )
    }
    
    func calculateHoldingCost(for vehicle: Vehicle, asOfDate: Date) -> Decimal {
        guard let settings = settings, settings.isEnabled else {
            return 0
        }
        
        guard Self.isHoldingCostEligible(vehicle: vehicle) else {
            return 0
        }
        
        let purchasePrice = vehicle.purchasePrice?.decimalValue ?? 0
        let expenses = vehicle.expenses?.allObjects as? [Expense] ?? []
        let baseExpenses = Self.getHoldingCostBaseExpenses(allExpenses: expenses)
        let expensesTotal = baseExpenses.reduce(Decimal(0)) { $0 + ($1.amount?.decimalValue ?? 0) }
        
        let capitalTiedUp = purchasePrice + expensesTotal
        let daysInInventory = Self.calculateDaysInInventory(vehicle: vehicle, asOfDate: asOfDate)
        
        let annualRate = settings.annualRatePercent?.decimalValue ?? 15.0
        let dailyRate = Self.calculateDailyRate(annualRatePercent: annualRate)
        
        return Decimal(daysInInventory) * dailyRate * capitalTiedUp
    }

    static func calculateDailyRate(annualRatePercent: Decimal) -> Decimal {
        guard annualRatePercent > 0 else { return 0 }
        return annualRatePercent / 365.0 / 100.0
    }
    
    static func calculateDaysInInventory(
        vehicle: Vehicle,
        asOfDate: Date = Date()
    ) -> Int {
        let purchaseDate = vehicle.purchaseDate ?? asOfDate
        let endDate = vehicle.saleDate ?? asOfDate
        let days = Calendar.current.dateComponents([.day], from: purchaseDate, to: endDate).day ?? 0
        return max(0, days)
    }
    
    static func getImprovementExpenses(allExpenses: [Expense]) -> [Expense] {
        allExpenses.filter { $0.deletedAt == nil && $0.categoryTypeEnum == .improvement }
    }
    
    static func getHoldingCostBaseExpenses(allExpenses: [Expense]) -> [Expense] {
        allExpenses.filter { $0.deletedAt == nil && $0.categoryTypeEnum != .holdingCost }
    }
    
    static func calculateDailyHoldingCost(
        vehicle: Vehicle,
        settings: HoldingCostSettings,
        improvementExpenses: [Expense]
    ) -> Decimal {
        guard settings.isEnabled else { return 0 }
        guard isHoldingCostEligible(vehicle: vehicle) else { return 0 }
        
        let purchasePrice = vehicle.purchasePrice?.decimalValue ?? 0
        let expensesTotal = improvementExpenses.reduce(Decimal(0)) { $0 + ($1.amount?.decimalValue ?? 0) }
        let capitalTiedUp = purchasePrice + expensesTotal
        
        let annualRate = settings.annualRatePercent?.decimalValue ?? 15.0
        let dailyRate = calculateDailyRate(annualRatePercent: annualRate)
        
        return capitalTiedUp * dailyRate
    }
    
    static func calculateAccumulatedHoldingCost(
        vehicle: Vehicle,
        settings: HoldingCostSettings,
        allExpenses: [Expense]
    ) -> Decimal {
        guard settings.isEnabled else { return 0 }
        guard isHoldingCostEligible(vehicle: vehicle) else { return 0 }
        
        let baseExpenses = getHoldingCostBaseExpenses(allExpenses: allExpenses)
        let dailyCost = calculateDailyHoldingCost(
            vehicle: vehicle,
            settings: settings,
            improvementExpenses: baseExpenses
        )
        
        let daysInInventory = calculateDaysInInventory(vehicle: vehicle)
        return Decimal(daysInInventory) * dailyCost
    }
    
    static func isHoldingCostEligible(vehicle: Vehicle) -> Bool {
        guard let status = vehicle.status, !status.isEmpty else { return true }
        return eligibleStatuses.contains(status)
    }
}

File: Android Car Dealer Tracker/app/src/main/java/com/ezcar24/business/util/calculator/HoldingCostCalculator.kt
package com.ezcar24.business.util.calculator

import com.ezcar24.business.data.local.Expense
import com.ezcar24.business.data.local.ExpenseCategoryType
import com.ezcar24.business.data.local.HoldingCostSettings
import com.ezcar24.business.data.local.Vehicle
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.Date
import java.util.concurrent.TimeUnit

object HoldingCostCalculator {

    private const val INTERMEDIATE_SCALE = 4
    private const val RATE_SCALE = 8
    private const val DISPLAY_SCALE = 2
    private const val DAYS_IN_YEAR = 365

    fun calculateDailyHoldingCost(
        vehicle: Vehicle,
        settings: HoldingCostSettings,
        improvementExpenses: List<Expense>
    ): BigDecimal {
        if (!settings.isEnabled) {
            return BigDecimal.ZERO.setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
        }

        val capitalTiedUp = getCapitalTiedUp(vehicle, improvementExpenses)
        val dailyRate = calculateDailyRateFromAnnual(settings.annualRatePercent)

        return capitalTiedUp
            .multiply(dailyRate)
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateAccumulatedHoldingCost(
        vehicle: Vehicle,
        settings: HoldingCostSettings,
        allExpenses: List<Expense>,
        asOfDate: Date = Date()
    ): BigDecimal {
        if (!settings.isEnabled) {
            return BigDecimal.ZERO.setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
        }

        val daysInInventory = calculateDaysInInventory(vehicle, asOfDate)
        if (daysInInventory <= 0) {
            return BigDecimal.ZERO.setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
        }

        val improvementExpenses = getImprovementExpenses(allExpenses)
        val dailyHoldingCost = calculateDailyHoldingCost(vehicle, settings, improvementExpenses)

        return dailyHoldingCost
            .multiply(BigDecimal(daysInInventory))
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateDaysInInventory(vehicle: Vehicle, asOfDate: Date = Date()): Int {
        val startDate = vehicle.purchaseDate
        val endDate = vehicle.saleDate ?: asOfDate

        val diffInMillis = endDate.time - startDate.time
        val days = TimeUnit.MILLISECONDS.toDays(diffInMillis).toInt()

        return maxOf(0, days)
    }

    fun getImprovementExpenses(allExpenses: List<Expense>): List<Expense> {
        return allExpenses.filter { expense ->
            expense.deletedAt == null && expense.expenseType == ExpenseCategoryType.IMPROVEMENT
        }
    }

    fun getCapitalTiedUp(vehicle: Vehicle, improvementExpenses: List<Expense>): BigDecimal {
        val improvementsTotal = improvementExpenses
            .filter { it.deletedAt == null }
            .map { it.amount }
            .fold(BigDecimal.ZERO) { acc, amount -> acc.add(amount) }

        return vehicle.purchasePrice
            .add(improvementsTotal)
            .setScale(INTERMEDIATE_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateDailyRateFromAnnual(annualRatePercent: BigDecimal): BigDecimal {
        return annualRatePercent
            .divide(BigDecimal(DAYS_IN_YEAR), RATE_SCALE, RoundingMode.HALF_UP)
            .divide(BigDecimal(100), RATE_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateAnnualRateFromDaily(dailyRatePercent: BigDecimal): BigDecimal {
        return dailyRatePercent
            .multiply(BigDecimal(DAYS_IN_YEAR))
            .multiply(BigDecimal(100))
            .setScale(INTERMEDIATE_SCALE, RoundingMode.HALF_UP)
    }
}
